Userland/Libraries/LibUnicode/URL.cpp

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2023, Simon Wanner <simon@skyrising.xyz>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <LibUnicode/IDNA.h>
+#include <LibUnicode/URL.h>
+
+namespace Unicode {
+
+// https://url.spec.whatwg.org/#concept-domain-to-ascii
+static ErrorOr<String> domain_to_ascii(StringView domain, bool be_strict)
+{
+    // 1. Let result be the result of running Unicode ToASCII with domain_name set to domain, UseSTD3ASCIIRules set to beStrict, CheckHyphens set to false, CheckBidi set to true, CheckJoiners set to true, Transitional_Processing set to false, and VerifyDnsLength set to beStrict. [UTS46]
+    // 2. If result is a failure value, domain-to-ASCII validation error, return failure.
+    auto result = TRY(Unicode::IDNA::to_ascii(
+        Utf8View(domain),
+        Unicode::IDNA::CheckHyphens::No,
+        Unicode::IDNA::CheckBidi::Yes,
+        Unicode::IDNA::CheckJoiners::Yes,
+        be_strict ? Unicode::IDNA::UseStd3AsciiRules::Yes : Unicode::IDNA::UseStd3AsciiRules::No,
+        Unicode::IDNA::TransitionalProcessing::No,
+        be_strict ? Unicode::IDNA::VerifyDnsLength::Yes : Unicode::IDNA::VerifyDnsLength::No));
+
+    // 3. If result is the empty string, domain-to-ASCII validation error, return failure.
+    if (result.is_empty())
+        return Error::from_string_literal("Empty domain");
+
+    // 4. Return result.
+    return result;
+}
+
+// https://url.spec.whatwg.org/#concept-host-parser
+URL create_unicode_url(String const& url_string)
+{
+    // NOTE: 1.-4. are implemented in URLParser::parse_host
+
+    URL url = url_string;
+    if (!url.is_valid() || url.host().is_null())
+        return url;
+
+    auto& domain = url.host();
+    // 5. Let asciiDomain be the result of running domain to ASCII with domain and false.
+    auto ascii_domain = domain_to_ascii(domain.view(), false);
+    // 6. If asciiDomain is failure, then return failure.
+    if (ascii_domain.is_error())
+        url.set_invalid();
```

Comment:
Sounds like we should be returning an Error here? (which then means we can just `TRY` the domain-to-ASCII step)

Reply:
This mirrors how `URLParser` handles errors, it returns a `URL` that's parsed as much as possible with the invalid flag set.

Reply:
I've been toying with the idea of removing `URL::is_valid()` entirely and using `ErrorOr` instead. I think if possible we want to move away from APIs that return something that's in an invalid state.

Reply:
Will just add to this by agreeing with Tim and Sam - I've been wanting to change `URLParser` to reject invalid URLs and be converted to `ErrorOr<URL>`. There are already some FIXME's lying around for this, so it would be good to not further extend the old pattern.